docs/usage.md

```diff
@@ -116,12 +147,141 @@ sudo podman run --rm -ti fzf
+```
+
+### Pre-fetch dependencies (pip)
+
+```shell
+cachi2 fetch-deps --source ./atomic-reactor '{
+  "type": "pip",
+  "requirements_files": ["requirements.txt"],
+  "requirements_build_files": ["requirements-build.txt", "requirements-pip.txt"]
+}'
+```
+
+Details: [pre-fetch dependencies](#pre-fetch-dependencies)
+
+### Generate environment variables (pip)
+
+```shell
+cachi2 generate-env ./cachi2-output -o ./cachi2.env --for-output-dir /tmp/cachi2-output
+```
+
+```shell
+$ cat cachi2.env
+export PIP_FIND_LINKS=/tmp/cachi2-output/deps/pip
+export PIP_NO_INDEX=true
+```
+
+Details: [generate environment variables](#generate-environment-variables)
+
+### Inject project files (pip)
+
+```shell
+$ cachi2 inject-files ./cachi2-output --for-output-dir /tmp/cachi2-output
+2023-01-26 16:41:09,990 INFO Overwriting /tmp/test/atomic-reactor/requirements.txt
+```
+
+The relevant part of the diff:
+
+```diff
+diff --git a/requirements.txt b/requirements.txt
+-osbs-client @ git+https://github.com/containerbuildsystem/osbs-client@8d7d7fadff38c8367796e6ac0b3516b65483db24
+-    # via -r requirements.in
++osbs-client @ file:///tmp/cachi2-output/deps/pip/github.com/containerbuildsystem/osbs-client/osbs-client-external-gitcommit-8d7d7fadff38c8367796e6ac0b3516b65483db24.tar.gz
+```
+
+Details: [inject project files](#inject-project-files)
+
+### Build the base image (pip)
+
+For this example, we will split the build into two parts - a base image and the final application image. In the base
+image build, we will cheat a bit and install "devel" libraries from RPMs. That means we won't be able to use network
+isolation (need to download the RPMs).
+
+If your project doesn't need to compile as many C packages as atomic-reactor, you may be able to find a base image that
+already contains everything you need.
+
+Containerfile.baseimage:
+
+```Dockerfile
+FROM quay.io/centos/centos:stream8
+
+# python3.8 runtime, C build dependencies
+RUN dnf -y install \
+        python38 \
+        python38-pip \
+        python38-devel \
+        gcc \
+        make \
+        libffi-devel \
+        krb5-devel \
+        cairo-devel \
+        cairo-gobject-devel \
+        gobject-introspection-devel \
+        openssl-devel && \
+    dnf clean all
+```
+
+Build the image:
+
+```shell
+podman build . -f Containerfile.baseimage --tag atomic-reactor-base-image:latest
+```
+
+### Build the container (pip)
+
+We will base the final application image on our custom base image. The base image build installed all the RPMs we will
+need, so the final phase can use network isolation again 🎉.
+
+Containerfile:
+
+```Dockerfile
+FROM atomic-reactor-base-image:latest
+
+COPY atomic-reactor/ /src/atomic-reactor
+WORKDIR /src/atomic-reactor
+
+# We're using network isolation => cannot build the cryptography package with Rust
+# (it downloads Rust crates)
+RUN export CRYPTOGRAPHY_DONT_BUILD_RUST=1 && \
+    source /tmp/cachi2.env && \
```

Comment:
Should we somehow emphasize we MUST `source /tmp/cachi2.env` and that it needs to be done in the same RUN instruction? An inline comment should be enough.

Reply:
Ack, done